lib/beaker/hypervisor/openstack.rb

```
@@ -250,83 +253,114 @@ def get_floating_ip
+    end
 
-      @hosts.each do |host|
-        if @options[:openstack_floating_ip]
-          ip = get_floating_ip
-          hostname = ip.ip.gsub('.','-')
-          host[:vmhostname] = hostname + '.rfc1918.puppetlabs.net'
-        else
-          hostname = ('a'..'z').to_a.shuffle[0, 10].join
-          host[:vmhostname] = hostname
-        end
-
-        create_or_associate_keypair(host, hostname)
-        @logger.debug "Provisioning #{host.name} (#{host[:vmhostname]})"
-        options = {
-          :flavor_ref => flavor(host[:flavor]).id,
-          :image_ref  => image(host[:image]).id,
-          :nics       => [ {'net_id' => network(@options[:openstack_network]).id } ],
-          :name       => host[:vmhostname],
-          :hostname   => host[:vmhostname],
-          :user_data  => host[:user_data] || "#cloud-config\nmanage_etc_hosts: true\n",
-          :key_name   => host[:keyname],
-        }
-        options[:security_groups] = security_groups(@options[:security_group]) unless @options[:security_group].nil?
-        vm = @compute_client.servers.create(options)
-
-        #wait for the new instance to start up
-        try = 1
-        attempts = @options[:timeout].to_i / SLEEPWAIT
-
-        while try <= attempts
+    # Parallel creation wrapper
+    def provision_parallel
+      # Array to store threads
+      @hosts.map do |host|
+        Thread.new do
           begin
-            vm.wait_for(5) { ready? }
-            break
-          rescue Fog::Errors::TimeoutError => e
-            if try >= attempts
-              @logger.debug "Failed to connect to new OpenStack instance #{host.name} (#{host[:vmhostname]})"
-              raise e
-            end
-            @logger.debug "Timeout connecting to instance #{host.name} (#{host[:vmhostname]}), trying again..."
+            create_instance_resources(host)
+          rescue => e
+            # Handle exceptions in the thread
+            puts "Thread #{host} failed with error: #{e.message}"
```

I feel like this ignore exceptions, and maybe this can be an issue when an exception is raised?

We can tell ruby to re-raise exception in the main thread when an exception occur and terminate a thread:
https://ruby-doc.org/3.3.0/Thread.html#method-c-abort_on_exception-3D

But for this to happen I think we also need to re-raise the exception with something like `raise e`.

I took the habit to always abort_on_exception when doing threads with Ruby, otherwise failures are sometimes hidden and we have unexpected behavior.

---

Yes this was to allow other threads to complete, but any failing ones would still constitute an acceptance test failure.  I have however modified it to now exit on a single failure

---

At the very least it should output to stderr, not stdout.

However, it does raise the question about cleanup. Does it tear down created hosts if one fails? So if Host A is created, but Host B fails, does it destroy host A again?

---

> At the very least it should output to stderr, not stdout.
> 
Agree, will test this and push any positive changes

> However, it does raise the question about cleanup. Does it tear down created hosts if one fails? So if Host A is created, but Host B fails, does it destroy host A again?

Good question, my expectation is:
a) any failed instances are left 'orphaned' ie not destroyed (this is normal and oos)
b) if the 'fail fast' option on 1 thread killing all is favoured, then who knows, probably they are all left orphaned
c) compreshensive scenario testing is i) out of scope and verging into the fog gem territory, and ii) a pita

On a successful run, instances are still destroyed as expected. :-)

---

I've done real-world testing and found that it is in fact best if all threads are terminated if one fails, due to the `hack_etc_hosts` function trying to run on failed hosts that have been cleaned up.  I have also invoked the `cleanup` function within the error handling.  I tested 4 permutations:

1. sequentially created hosts with valid params for the gem to pass (completes and cleans up all threads/hosts unless `BEAKER_destroy=no` is set)
2. sequentially created hosts with an invalid param (flavor) for the gem to pass (aborts on encountering error and cleans up all threads/hosts)
3. parallel created hosts with valid params for the gem to pass (completes faster and cleans up all threads/hosts unless `BEAKER_destroy=no` is set)
4. parallel created hosts with an invalid param (flavor) for the gem to pass (aborts very fast on error and cleans up all threads/hosts)

Also, I changed `$stderr` message to use `@logger.error` as per convention elsewhere in the file.

I'm happy with this functionality after conducting real-world testing.

Main changes to do this are:

``` 
    # Handle exceptions in the thread
    rescue => e
      @logger.error "Thread #{host} failed with error: #{e.message}"
      # Call cleanup function to prevent orphaned hosts
      cleanup
      # Pass the error to the main thread to terminate all threads
      Thread.main.raise(e)
      # Terminate the current thread (to prevent hack_etc_hosts trying to run)
      Thread.kill(Thread.current)

      hack_etc_hosts @hosts, @options
```